### ISSUE TITLE
Truffle Interop Types: Array & Vectors

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#4542]
This task is automatically imported from the old Task Issue Board and it was originally created by jaroslavtulach.
Original issue is [here](https://www.pivotaltracker.com/story/show/183000876).
<hr />

Truffle polyglot system allows various types to be exchanged between different languages (Enso, JavaScript, Python, but also calls to hosted Java). These types include JVM primitive types, strings, date & time values as well as arrays: https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/InteropLibrary.html

The handling of these types in Enso shall be uniform, based on the [InteropLibrary](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/InteropLibrary.html) and should be opaque - e.g. not piggyback on the knowledge of the origin of the value. Rather it shall use just the [InteropLibrary](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/interop/InteropLibrary.html) messages and treat all the values the same. If it _“quacks like a date”_, it is a date…

Another principle is as late conversion as possible. Truffle language shall let the foreign value flow through as long as possible to minimize the cost of conversion, hoping the conversion may not be needed at all. On the other hand, it is necessary to make sure that the conversion of Enso non-supported types (`byte`, `int`, `float`) happens before they propagate to the internal AST of Enso interpreter.

### From Polyglot to Enso

An example of this approach has been demonstrated in https://github.com/enso-org/enso/pull/3559 and it consists primarily of:
* doConvertDate - specialization to invoke messages on the object
* PolyglotProxy delegation of necessary messages
* getPolyglotCallType check
* EnsoDate wrapper representing the date internally in Enso ASTs

The implementation of Array support shall follow the same:
* doConvertArray check if the foreign object supports hasArrayElements message
* PolygotProxy delegation to all the array related messages (probably done already)
* getPolyglotCallType check to detect hasArrayElements support
* Array builtin wrapper that delegates all its operations to Object using InteropLibrary
* Conversion of getArrayElement values to Enso internally supported ones

That way we can have uniform, opaque support for polyglot arrays in Enso.
Everything that _“ducks like an array”_ will be treated as array
* All such objects will have Enso `Array` methods
* No other methods will (need to) be exposed
* Just one Enso Array type in the whole system
* Yet there can be multiple `XYZArray` **extends** `TruffleObject` floating in the interpreter

### From Enso to Polyglot

There are two types in Enso that represent a (polyglot) array:
* `Array` - wrapper around (potentially mutable) foreign array
* `Vector` - native Enso immutable array

Both these types may want to participate in polyglot interop with Java or other languages. To simplify conversion of Enso native data structures to other languages, both of these types shall be implemented as builtin types (just like `EnsoDate`) and both shall_ “quack like an array”_ - e.g. support `hasArrayElements` and `getArrayElement` messages.

_Implementation note_: Theoretically both Enso types (`Array` and `Vector`) could be implemented by the same `TruffleObject` subclass - they are not going to be mutable from outside, they’d rather have the same methods and they should look the same from a perspective of other languages. The only difference is that `Array` represents something that can potentially mutate, but `Vector` guarantees that its content is going to stay the same.

### From Java to Enso without Copying

If there is a Java implementation of a table which represents a huge array - can it be used as Enso `Array` (or `Vector`) without any copying? Yes, enough to implement List or directly polyglot SDK `ProxyArray` - that way it is going to be perceived as `hasArrayElements` object - e.g. _“duck like an array”_ and will be automatically represented as Enso `Array`.

### Test Cases

Create an array in JavaScript, Python and return it to Enso. All methods of `Array` shall be available instantly.

Invoke a Java method that returns `Object[]` and use it from Enso. All methods of Array shall be available instantly.

Invoke a Java method that calls back and passes `Object[]` as an argument. Use the array from Enso. All methods of `Array` shall be available instantly.

Create an Enso `Array` with `Date` instances. Pass the array to Java method that accepts `List<LocalDate>` - verify dates are properly available.

Repeat the previous `Array` exercise with `Vector`. The result shall be the same.

Invoke a Java method that creates an array with all primitive types, `String`, date and _another array_ of the same content. Make sure that it all gets converted to appropriate Enso types - e.g. `Array` of `long` and `Text` and another `Array`.

Create Java method that returns `List` - call it from Enso to get `Array`. Invoke length check and verify only the size() method was called - items of the array haven’t been touched.

Repeat the previous exercise with `ProxyArray`. Again, no elements shall be created.

### Prior Work

Originally described [at google doc](https://docs.google.com/document/d/1nrOBQeCf139e5Hav5zgvSwWT0OJ2MYuIk_2b8UZ3sDo/).

## Tasks:
- [X] Create a builtin type for Array
- [X] Make sure all its (core) methods operate via InteropLibrary
- [X] Implement conversions of everything that _ducks like an array_ to such type
- [X] Make vector a builtin as well
- [X] Make sure `Vector` and `Array` share the same set of methods - reported as new [user story](https://www.pivotaltracker.com/n/projects/2539304/stories/183217534)
- [X] Convert values obtained from an `Array` to proper Enso types
- [X] Can a List/ProxyArray value be passed to Enso without copying?
- [X] Run & compare benchmark results

## Comments:
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-08-22):

**Progress:** Started work towards improved Vector/Array. In the meantime fought with CI to get the interim #3628 PR in. It should be finished by 2022-08-24.

**Next Day:** Next day I will be working on the [##183000876 task](https://www.pivotaltracker.com/story/show/183000876). Pick up the pending PRs (waiting on review of time PR) and continue investigation into Arrays. (Enso Bot - Aug 23, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-08-23):

**Progress:** Slowly progressing on builtin Vector. Addressed feedback for Date_Time PR. Slightly distracted by various random CI failures. It should be finished by 2022-08-24.

**Next Day:** Next day I will be working on the [##183000876 task](https://www.pivotaltracker.com/story/show/183000876). Hopefully merge Date_Time PR tomorrow. Prepare WIP for current work on Vector so that Jaroslav can pick it up. (Enso Bot - Aug 24, 2022)
<hr />
**Hubert Plociniczak** reports a new **STANDUP** for yesterday (2022-08-24):

**Progress:** Good progress builtin Vector. Almost all `test/Test` passes. Fixing a few remaining ones will be straightforward. WIP PR is up for anyone to pick up. It should be finished by 2022-08-24.

**Next Day:** Next day I will be working on the [##183000876 task](https://www.pivotaltracker.com/story/show/183000876). Either continue on Vector/Array interop or whatever is next in line by then. (Enso Bot - Aug 24, 2022)
<hr />
**Jaroslav Tulach** reports a new **🔴 DELAY** for yesterday (2022-08-29):

**Summary:** There is **7 days** delay in implementation of the [Truffle Interop Types: Array & Vectors (#183000876) task](https://www.pivotaltracker.com/story/show/183000876).
It **will cause 7 days delay** for the delivery of this weekly plan.

I've just took over the issue yesterday; moreover Marcin's current work has merge priority

**Delay Cause:** The issue was scheduled to be addressed this week

**Possible solutions:** Address the issues in the PR and integrate once Marcin's work is done (Enso Bot - Aug 30, 2022)
<hr />
**Jaroslav Tulach** reports a new **🔴 DELAY** for yesterday (2022-08-29):

**Summary:** There is **7 days** delay in implementation of the [Truffle Interop Types: Array & Vectors (#183000876) task](https://www.pivotaltracker.com/story/show/183000876).
It **will cause 7 days delay** for the delivery of this weekly plan.

I've just took over the issue yesterday; moreover Marcin's current work has merge priority

**Delay Cause:** The issue was scheduled to be addressed this week

**Possible solutions:** Address the issues in the PR and integrate once Marcin's work is done (Enso Bot - Aug 30, 2022)
<hr />
**Jaroslav Tulach** reports a new **🔴 DELAY** for yesterday (2022-08-29):

**Summary:** There is **12 days** delay in implementation of the [Truffle Interop Types: Array & Vectors (#183000876) task](https://www.pivotaltracker.com/story/show/183000876).
It **will cause 12 days delay** for the delivery of this weekly plan.

I've just took over the issue yesterday; moreover Marcin's current work has merge priority

**Delay Cause:** The issue was scheduled to be addressed this week

**Possible solutions:** Address the issues in the PR and integrate once Marcin's work is done (Enso Bot - Aug 30, 2022)
<hr />
**Jaroslav Tulach** reports a new **STANDUP** for yesterday (2022-08-29):

**Progress:** Merge of improved text visualization #3665.Array and Vector work: `VectorLiteralsTest`: https://github.com/enso-org/enso/pull/3667/commits/40cccac69ae9f58d96e3b6e75160b240d9fc1dea; Debugging Number.should_equals: https://github.com/enso-org/enso/pull/3667/commits/c42663932a54621106cb8eec239c8113500143e5; Char conversion: https://github.com/enso-org/enso/pull/3667/commits/465a269408cea20d026bfb9c825e0803a491178d; Debugging should capture traces correctly test; Merging with develop and removal of set_at; Fighting with should correctly propagate error through map: https://github.com/enso-org/enso/pull/3667/commits/8c6a1cc138270492bc495aefa35bc4d653f9add0; Fixing tables: https://github.com/enso-org/enso/pull/3667/commits/cf6a72a21bdfebd44a21f017927709dfb7abf626; Meetings, demos, lazy visualizations It should be finished by 2022-09-19.

**Next Day:** Continue Array & Vector work https://github.com/enso-org/enso/pull/3667; plan lazy visualizations (Enso Bot - Aug 30, 2022)
<hr />
**Jaroslav Tulach** reports a new **STANDUP** for yesterday (2022-08-30):

**Progress:** `Array` and `Vector` and lazy visualizations It should be finished by 2022-09-19.

**Next Day:** Fix CI gates for Array & Vector work https://github.com/enso-org/enso/pull/3667 (Enso Bot - Aug 31, 2022)
<hr />
**Jaroslav Tulach** reports a new **STANDUP** for yesterday (2022-08-31):

**Progress:** Merged Marcin's work with Array and Vector. Bugfixed everything (often `case of`: https://github.com/enso-org/enso/pull/3667/commits/aa17c2bbf39f8bd56e9c4f67acd86a8c52730a45 and https://github.com/enso-org/enso/pull/3667/commits/7df85d80aa1afc07089e5fb06296394526b14001). Meetings. It should be finished by 2022-09-19.

**Next Day:** Fix last five(?) failures in `Table_Tests`. Pass CI with Array and Vector work. (Enso Bot - Sep 1, 2022)
<hr />
**Jaroslav Tulach** reports a new **STANDUP** for yesterday (2022-09-01):

**Progress:** All CI checks passed for `Vector` and `Array` as of https://github.com/enso-org/enso/pull/3667#issuecomment-1233833584 - requested review. Adding tests to cover more interop (Java, JavaScript, Python) aspects. It should be finished by 2022-09-19.

**Next Day:** More tests and fixes for the _polyglot interop_ of `Vector` and `Array` (Enso Bot - Sep 2, 2022)
<hr />
**Jaroslav Tulach** reports a new **STANDUP** for yesterday (2022-09-02):

**Progress:** Enso Array methods rewritten to work on any array-like object - [dispatch modified](https://github.com/enso-org/enso/pull/3667#issuecomment-1235710717). Trying to compare `Array` and `Vector` in a `Test` suite. Defining `Array.==`. Trying coercion in `Vector`. It should be finished by 2022-09-19.

**Next Day:** Address the current ~10 failures in CI of `Array` & `Vector` (Enso Bot - Sep 3, 2022)
<hr />
**Jaroslav Tulach** reports a new **STANDUP** for yesterday (2022-09-05):

**Progress:** Proposed an [umbrella for drop-downs PR](https://github.com/enso-org/design/pull/41); Reported slowness https://www.pivotaltracker.com/n/projects/2539304/stories/183176528 - possibly only caused by Vector & Array work: https://github.com/enso-org/enso/pull/3667/commits/542d85f37823b46743a7b10b9cd1f9e2b461f7a8; Updated parser branch to latest `develop`: https://github.com/enso-org/enso/pull/3611 It should be finished by 2022-09-19.

**Next Day:** Vector & Array CI cleanup and parser cleanup (Enso Bot - Sep 6, 2022)
<hr />
**Jaroslav Tulach** reports a new **STANDUP** for yesterday (2022-09-06):

**Progress:** Updating parser in the morning: parse type with data definitions: https://github.com/enso-org/enso/pull/3611/commits/a5f74a605ee46585ff8821fdc55c6d49197ee296 & parsing `export`. `Vector` and `Array` performance, trying to _fix_ CI. It should be finished by 2022-09-19.

**Next Day:** Ask for review Vector & Array. (Enso Bot - Sep 7, 2022)
<hr />
Another task fixed:

> Check behavior of copy/wrap over List<String> - 8336c8e
> Check the lazy behavior for `ProxyArray` - 981829c  (jaroslavtulach - Sep 7, 2022)
<hr />
The first picture shows _benchmark results produced by CI_ on Sep 8, 2022. Executed on [develop branch](https://github.com/enso-org/enso/actions/runs/3007625706) and against [PR-3667 branch](https://github.com/enso-org/enso/actions/runs/3010582834). Am I just dreaming or did we manage to speed up Enso in all the benchmarks? 

As the _CI measurements seem suspicious_ I've run the same measurements on my own computer over night. There is no improvement, but also no regression in the second picture. Good from _Array & Vector_ point of work, not really good from the _CI infrastructure_ point of view. CCing @murbanczyk .

I've just scheduled final CI test on Sep 12-13: [develop](https://github.com/enso-org/enso/actions/runs/3039071636) vs. [wip/hubert/array-vector-interop-183000876](https://github.com/enso-org/enso/actions/runs/3042020612). These two measurement confirm the manual ones - _no regression, no improvement_.
 (jaroslavtulach - Sep 7, 2022)
<hr />
**Jaroslav Tulach** reports a new **STANDUP** for yesterday (2022-09-07):

**Progress:** Test and fix for `Atom.getMembers`: https://github.com/enso-org/enso/pull/3689; Defensively copy in `Vector.from_array`: https://github.com/enso-org/enso/pull/3667/commits/f628b28f5ffa0bcdceb64a7604317b2f7aee58bd; Meeting about `BigInteger`: https://www.pivotaltracker.com/story/show/182962982/comments/233179278; Using cached interop for `Vector`: https://github.com/enso-org/enso/pull/3667/commits/f0338c5e1a2a2621db4cf4cbc802a7b0d746cca5; test for lazy List: https://github.com/enso-org/enso/pull/3667/commits/8336c8ea6ae9a5ef0143c7178f7ccd3f4c5e109c; test for proxy array: https://github.com/enso-org/enso/pull/3667/commits/981829c38ab4e64959db547c6abf13c8f673172a; using `from_polyglot_array`: https://github.com/enso-org/enso/pull/3667/commits/ce6aa5cddb9e6aa3cb1a3f46510971bd34733b33; benchmarks results look very positive: https://www.pivotaltracker.com/story/show/183000876/comments/233194239 It should be finished by 2022-09-19.

**Next Day:** Address review comments for Vector & Array. (Enso Bot - Sep 7, 2022)
<hr />
**Jaroslav Tulach** reports a new **STANDUP** for yesterday (2022-09-08):

**Progress:** Proposing solution for broken IR caches: https://github.com/enso-org/enso/pull/3692; Meeting about Enso/Java StdLib interop: https://docs.google.com/document/d/1XOnkBJjZuiAhIwjtekSa8LjHtGLDHqGwyuEanei43wg; Pair debugging Truffle source code; Found (_very useful_) `ReflectiveLibrary`: https://github.com/enso-org/enso/pull/3667/commits/f6340361a0d82047c56b59f8f2d20a209ba6c0b7#diff-d784205295fc228b108d6ca365bf8371e6fa2337bd8f2c38e7a3bdac721f3a30R14; Some engine CI gates are green for _Vector & Array_: https://github.com/enso-org/enso/commit/ef71c207342ca819aaf89ece2cb6f2a94c2a017d It should be finished by 2022-09-19.

**Next Day:** Write benchmarks for Vector & Array. (Enso Bot - Sep 9, 2022)
<hr />
**Jaroslav Tulach** reports a new **STANDUP** for yesterday (2022-09-09):

**Progress:** CI fighting: https://discord.com/channels/401396655599124480/1016968486754471966/1017642222558130186; but [Broken IR caches fix](https://github.com/enso-org/enso/pull/3692) integrated; `toDisplayString` for `Vector`: https://github.com/enso-org/enso/pull/3667/commits/aaf33e3d357cb2364deac31cd0d63293833e2dae; [Comparative benchmark](https://github.com/enso-org/enso/pull/3667/commits/96ad6f4210d9b059dafe83fe2c7db111b0f95c50) for _Vector&Array_ and [fix to make it fast](https://github.com/enso-org/enso/pull/3667/commits/a3b9b6e8bcb4fd4e6d59f236849703dfbcac0d25); Preparing presentation for GraalVM engineers for next week's _GraalVM meetup_ - [slides 12-19](https://docs.google.com/presentation/d/1oyUdGKSrTA-KpEjScPfFcNzkBK9jX4YEobp3whHOr5Y) It should be finished by 2022-09-19.

**Next Day:** Prepare for the GraalVM meetup (investigate Espresso, native-image). (Enso Bot - Sep 10, 2022)
<hr />
**Jaroslav Tulach** reports a new **STANDUP** for yesterday (2022-09-10):

**Progress:** Polishing _Vector & Array_: https://github.com/enso-org/enso/pull/3667 - it is green now. Reporting few TODOs into pivotal for the engine.Reported https://github.com/oracle/graal/pull/4916 to GraalVM team It should be finished by 2022-09-19.

**Next Day:** Prepare for the GraalVM meetup (investigate Espresso, native-image). (Enso Bot - Sep 11, 2022)
<hr />
**Jaroslav Tulach** reports a new **STANDUP** for yesterday (2022-09-12):

**Progress:** Preparing for GraalVM meetup. Testing native image - backport of bugfix on GraalVM side needed: https://github.com/oracle/graal/issues/4200; providing native image config files: https://github.com/enso-org/enso/pull/3638/commits/b4292d3120ed5ac640618ac77accbde8b9ab5ec9; PR to simplify _hello world_ demo in Enso: https://github.com/enso-org/enso/pull/3696; Meetings. Listing technical debts in engine: https://discord.com/channels/401396655599124480/1018889551441313894/1018891614673973278 It should be finished by 2022-09-19.

**Next Day:** Prepare for the GraalVM meetup (investigate Espresso). (Enso Bot - Sep 13, 2022)
<hr />
**Jaroslav Tulach** reports a new **STANDUP** for yesterday (2022-09-13):

**Progress:** Finally merged _vector & array_: https://github.com/enso-org/enso/pull/3667;
Demo of using Espresso for Java interop: https://www.pivotaltracker.com/n/projects/2539304/stories/183260380;
Discussions about Espresso interop with GraalVM team, bugreport: https://github.com/oracle/graal/issues/4939;
Discussing access to Java properties of `Date`: https://docs.google.com/document/d/1ErTnBtX52egVietPd8O-vS_pH_yAB968BTvpQZbFrYE It should be finished by 2022-09-19.

**Next Day:** Traveling to Zurich (Enso Bot - Sep 14, 2022)
<hr />
**Jaroslav Tulach** reports a new **STANDUP** for yesterday (2022-09-14):

**Progress:** Travel to Zurich; Reading emails; Preparing patch for GraalVM for discussion: https://github.com/oracle/graal/pull/4916 It should be finished by 2022-09-19.

**Next Day:** GraalVM meetup, day 1 (Enso Bot - Sep 15, 2022)
<hr />
**Jaroslav Tulach** reports a new **STANDUP** for yesterday (2022-09-15):

**Progress:** GraalVM meetup, day 1 It should be finished by 2022-09-19.

**Next Day:** GraalVM meetup, day 2: Presentation about Enso (Enso Bot - Sep 16, 2022)
<hr />
**Jaroslav Tulach** reports a new **STANDUP** for yesterday (2022-09-16):

**Progress:** GraalVM meetup, Enso presentation: https://twitter.com/wojciech_danilo/status/1570719810409103361?s=20&t=R35qiDsU40vQJhHnQ3U7CA; travel back to Prague It should be finished by 2022-09-19.

**Next Day:** Back to coding! (Enso Bot - Sep 17, 2022)
<hr />
